### PR TITLE
Break out embedded lang classification from semantic classification.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
This makes embedded language classification (e.g. Regex/Json) independent from normal semantic classification.  This has the benefit of making sure that our semantic classification is not gated behind embedded language classification happening.

This is also important for future work as we want to expose embedded lang classification to other parties (like ASP to start with).  And we want to ensure that the classification work they do does not impact teh highly latency critical area of normal semantic classification.

PR should be viewed with whitespace off.

### CODE REVIEW

File: src/EditorFeatures/Core/Implementation/Classification/ClassificationUtilities.cs [hunk 1]
```diff
@@ -1,33 +0,0 @@
...
-        public static List<ITagSpan<IClassificationTag>> Convert(IClassificationTypeMap typeMap, ITextSnapshot snapshot, ArrayBuilder<ClassifiedSpan> classifiedSpans)
-        {
-            var result = new List<ITagSpan<IClassificationTag>>(capacity: classifiedSpans.Count);
-            foreach (var span in classifiedSpans)
-                result.Add(Convert(typeMap, snapshot, span));
-
-            return result;
-        }
-    }
-}
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/Classification/ClassificationUtilities.cs:33]: this merged with SemanticClassificationUtilities.

File: src/EditorFeatures/Core/Implementation/Classification/Semantic/ClassificationUtilities.cs [hunk 2]
```diff
@@ -26,14 +25,31 @@
...
+        }
+
+        public static List<ITagSpan<IClassificationTag>> Convert(IClassificationTypeMap typeMap, ITextSnapshot snapshot, ArrayBuilder<ClassifiedSpan> classifiedSpans)
+        {
+            var result = new List<ITagSpan<IClassificationTag>>(capacity: classifiedSpans.Count);
+            foreach (var span in classifiedSpans)
+                result.Add(Convert(typeMap, snapshot, span));
+
+            return result;
+        }
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/Classification/Semantic/ClassificationUtilities.cs:44]: these methods just moved when i merged tehse two utilities files.

File: src/EditorFeatures/Core/Implementation/Classification/Semantic/ClassificationUtilities.cs [hunk 3]
```diff
@@ -63,15 +78,17 @@ internal static class SemanticClassificationUtilities
...
             await ClassifySpansAsync(
-                context, spanToTag, classificationService, typeMap, options, cancellationToken).ConfigureAwait(false);
+                context, document, spanToTag.SnapshotSpan, classificationService, typeMap, options, type, cancellationToken).ConfigureAwait(false);
         }
 
         private static async Task<bool> TryClassifyContainingMemberSpanAsync(
             TaggerContext<IClassificationTag> context,
-            DocumentSnapshotSpan spanToTag,
+            Document document,
+            SnapshotSpan snapshotSpan,
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/Classification/Semantic/ClassificationUtilities.cs:87]: i broke this into two parameters in some cases as DocumentSnapsthoSpan has a nullable Document, and that makes the NRT analysis annoying.

File: src/EditorFeatures/Core/Tagging/TaggerContext.cs [hunk 4]
```diff
@@ -22,7 +22,7 @@ internal class TaggerContext<TTag> where TTag : ITag
     {
         private readonly ImmutableDictionary<ITextBuffer, TagSpanIntervalTree<TTag>> _existingTags;
 
-        internal IEnumerable<DocumentSnapshotSpan> _spansTagged;
+        internal ImmutableArray<SnapshotSpan> _spansTagged;
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Tagging/TaggerContext.cs:25]: _spansTagged never made use of the 'Document' portoin of hte span.  And the Document part was scary (to me) as we mutate the document when computing these (e.g. we get teh FrozenPartial snapshot.  i really didn't like accidentally having that live and be out of sync with the actual Doc being classified).

File: src/Workspaces/CSharp/Portable/Classification/SyntaxClassification/CSharpEmbeddedLanguageClassificationServiceFactory.cs [hunk 5]
```diff
@@ -5,22 +5,23 @@
...
         [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
-        public CSharpSyntaxClassificationServiceFactory()
+        public CSharpEmbeddedLanguageClassificationServiceFactory()
         {
         }
 
-        [Obsolete(MefConstruction.FactoryMethodMessage, error: true)]
         public ILanguageService CreateLanguageService(HostLanguageServices languageServices)
-            => new CSharpSyntaxClassificationService(languageServices);
+            => new EmbeddedLanguageClassificationService(languageServices.GetRequiredService<IEmbeddedLanguagesProvider>());
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/Classification/SyntaxClassification/CSharpEmbeddedLanguageClassificationServiceFactory.cs:25]: this is actually an entirely new type.  git decided to merge it with another file i deleted :'(

File: src/EditorFeatures/Core/Implementation/Classification/Semantic/AbstractSemanticOrEmbeddedClassificationViewTaggerProvider.cs [hunk 6]
```diff
@@ -32,35 +29,33 @@ namespace Microsoft.CodeAnalysis.Editor.Implementation.Classification
...
     /// </summary>
-    [Export(typeof(IViewTaggerProvider))]
-    [TagType(typeof(IClassificationTag))]
-    [ContentType(ContentTypeNames.RoslynContentType)]
-    internal partial class SemanticClassificationViewTaggerProvider : AsynchronousViewTaggerProvider<IClassificationTag>
+    internal abstract class AbstractSemanticOrEmbeddedClassificationViewTaggerProvider : AsynchronousViewTaggerProvider<IClassificationTag>
     {
         private readonly ClassificationTypeMap _typeMap;
         private readonly IGlobalOptionService _globalOptions;
+        private readonly ClassificationType _type;
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/Classification/Semantic/AbstractSemanticOrEmbeddedClassificationViewTaggerProvider.cs:36]: this is the crux of the change.  there are now two taggers.  they operate independently, but almost identically.  The only difference is the type of classifications they request, which is specified by this type.

**Review by davidwengier — APPROVED** [review]
Makes sense

**Pull request merged**